receiver/awsfirehosereceiver/internal/unmarshaler/otlpmetricstream/unmarshaler.go

```diff
@@ -4,73 +4,57 @@
 	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/extension"
 	"go.opentelemetry.io/collector/pdata/pmetric"
-	"go.opentelemetry.io/collector/pdata/pmetric/pmetricotlp"
-	"go.uber.org/zap"
 
-	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/awsfirehosereceiver/internal/metadata"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/extension/encoding/awscloudwatchmetricstreamsencodingextension"
 )
 
 const (
-	// Supported version depends on version of go.opentelemetry.io/collector/pdata/pmetric/pmetricotlp dependency
+	// TypeStr Supported version depends on version of go.opentelemetry.io/collector/pdata/pmetric/pmetricotlp dependency
 	TypeStr = "otlp_v1"
-)
 
-var errInvalidOTLPFormatStart = errors.New("unable to decode data length from message")
+	// format of the aws cloudwatch metric stream encoding extension
+	format = "opentelemetry1.0"
+)
 
 // Unmarshaler for the CloudWatch Metric Stream OpenTelemetry record format.
 //
 // More details can be found at:
 // https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/CloudWatch-metric-streams-formats-opentelemetry-100.html
 type Unmarshaler struct {
-	logger    *zap.Logger
-	buildInfo component.BuildInfo
+	unmarshaler pmetric.Unmarshaler
 }
 
 var _ pmetric.Unmarshaler = (*Unmarshaler)(nil)
 
-// NewUnmarshaler creates a new instance of the Unmarshaler.
-func NewUnmarshaler(logger *zap.Logger, buildInfo component.BuildInfo) *Unmarshaler {
-	return &Unmarshaler{logger, buildInfo}
-}
-
-// UnmarshalMetrics deserializes the recordsas a length-delimited sequence of
-// OTLP metrics into pmetric.Metrics.
-func (u Unmarshaler) UnmarshalMetrics(record []byte) (pmetric.Metrics, error) {
-	md := pmetric.NewMetrics()
-	dataLen, pos := len(record), 0
-	for pos < dataLen {
-		n, nLen := proto.DecodeVarint(record)
-		if nLen == 0 && n == 0 {
-			return md, errInvalidOTLPFormatStart
-		}
-		req := pmetricotlp.NewExportRequest()
-		pos += nLen
-		err := req.UnmarshalProto(record[pos : pos+int(n)])
-		pos += int(n)
-		if err != nil {
-			return pmetric.Metrics{}, fmt.Errorf("unable to unmarshal input: %w", err)
-		}
-		for i := 0; i < req.Metrics().ResourceMetrics().Len(); i++ {
-			rm := req.Metrics().ResourceMetrics().At(i)
-			for j := 0; j < rm.ScopeMetrics().Len(); j++ {
-				sm := rm.ScopeMetrics().At(j)
-				sm.Scope().SetName(metadata.ScopeName)
-				sm.Scope().SetVersion(u.buildInfo.Version)
-			}
-		}
-		req.Metrics().ResourceMetrics().MoveAndAppendTo(md.ResourceMetrics())
+// NewUnmarshaler creates a new instance of the aws cloudwatch metric streams
+// encoding extension unmarshaler for the OpenTelemetry record format.
+func NewUnmarshaler(ctx context.Context) (pmetric.Unmarshaler, error) {
+	f := awscloudwatchmetricstreamsencodingextension.NewFactory()
+	extensionID := component.NewID(f.Type())
+	ext, err := f.Create(ctx, extension.Settings{
+		ID: extensionID,
+	}, &awscloudwatchmetricstreamsencodingextension.Config{
```

Comment:
This means there's no telemetry settings or BuildInfo being passed in right? Doesn't seem ideal. I think we should create the extension in the receiver code, and delete internal/unmarshaler/otlpmetricstream.

Reply:
We have to change the `BuildInfo` to not use the scope name from the receiver. But since the scope name is coming from the internal package of the encoding extension, we would have to hardcode it here. WDYT?

Reply:
Sorry @axw, I was wrong!! The scope name is correct, since we define it directly on the unmarshaler from the `metadata` there.

Reply:
https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/58d4ce9f27658578f202c966d0d19f3b4e73a58e/extension/encoding/awscloudwatchmetricstreamsencodingextension/opentelemetry_unmarshaler.go#L54-L55

Reply:
Ah yes :) No worries, thanks for clarifying.